### DISCUSSION TITLE
Open collective project and goals: Bug bounties

### DISCUSSION THREAD

**Discussion by Aggror** [stride3d/stride#1204] [Contributors]
We can add another project to the open collective that uses part of the funding to set bounties on high priority bugs.
Once there are some good suggestion for bounties, the open collective project will be created and this discussion will be updated.

If you have any ideas for high priority bugs that you think deserves a bounty, list it here below. 

**Work in progress:**
**General rules for bug bounties:**
    - Before starting on an issue, make it known on Github and Discord that your are willing to tackle the issue. It would be a shame if multiple people try to solve the same issue.
    - Any PR to the Stride repo must be reviewed by a contributor and discussed if needed.
    - Once a PR is reviewed and merged to the Stride repo, the submitter 60% of the bounty.
    - The submitter receives the remaining 40% when stride release happens.

**Comment by Eideren** [comment]
Here are a couple I found:
#455
#615
#721
#1024

**Reply by Aggror** [reply]
Sorry haven't had the time to look at these (and kind a fell of my radar :)). Will try to have a look at these this weekend.

**Reply by Aggror** [reply]
I had a look at the bugs you posted @Eideren. Honestly I can't tell how difficult any of these problems are. I am just going to suggest some things, 

https://github.com/stride3d/stride/issues/455: UI Slowness
Requires extensive testing and quite bit of knowledge on the UI system. I don't know if Stride's UI system is dependent on a possible switch to Avalonia or MAUI in the future. If not, I would suggest putting up a bounty for $200 for investigation in the UI system and come up with a possible explanation and fix.

https://github.com/stride3d/stride/issues/615: Editor memory leak
How urgent would this be? In worst case scenario the editor crashes after a while of being used and you would have to restart. 

https://github.com/stride3d/stride/issues/1024 Vulkan full screen
I would suggest putting up a bounty for $200 for investigation in the vulkan system and come up with a possible explanation and fix.

**Reply by Eideren** [reply]
#455 that's the runtime UI. It shouldn't be built on top of those huge retained UIs, it's just not the right tool for the job, so I would say that it doesn't rely on the editor ui changes.

#615 is not that important, given that it's caused by a an exception, and should be fairly easy to fix anyway. You're right.

What about #1024 ? I think the user documented it pretty extensively, I'm sure assigning the money for fixing the issue directly would be fine ?

**Reply by Aggror** [reply]
So for the UI we could look in to a new UI framework? Or even write our own inname UI system?

As for the vulcan one goes, that is a prime candidate for a custom bug bounty

**Reply by Eideren** [reply]
@Aggror I'm not sure I follow, which UI are you referring to, in-game or editor ? For the in-game one I don't think I heard anyone talk about replacing it, if you want to talk about this best to do it on discord and report that specific issue for later then ?

**Reply by Aggror** [reply]
My bad, I thought that was what you meant with 'It shouldn't be built on top of those huge retained UIs, it's just not the right tool for the job, so I would say that it doesn't rely on the editor ui changes.'   

But we have 2 bugs bounties now that we can propose during the community meeting. We can make it official then. Thanks again for skimming over the large list of bugs!

**Comment by najak3d** [comment]
I'd like to make a strong recommendation for creating a Project and Bounty for "creating a Stride3D MAUI render Control for the Code-Only approach".    MAUI has ZERO solutions for this right now, and I'd like to see Stride3D become the first and best.

If someone will create this project and set a bounty for it -- I think I can raise the money for that bounty.   I will promote this in the MAUI communities which are comprised of mostly business apps, some of which would benefit from use of Stride3D.

How much does the bounty need to be, for this to get prioritized and done within a couple months?   I'm thinking $1000.

A similar integration was done using UrhoSharp in 2019 for Xamarin Forms and .NET Core.   But UrhoSharp is not maintained, nor can anyone figure out how to build it again -- so it's forever dormant, and functionally dead now.   The path is clear for Stride3D to now fill this void.

If someone on the team can spec out this project and create the bounty -- I should be able to quickly raise this money to pay this bounty, and get it done sooner rather than later.

**Reply by tebjan** [reply]
this is already announced here, be we first need to find developers to work on it: https://opencollective.com/stride3d/projects/embed-stride-ui-frameworks

**Reply by najak3d** [reply]
Thank you for showing me that.

We specifically want/need MAUI.   We have a GPS map app with 3D view that would benefit from Stride3D, vs. UrhoSharp.   We won't be able to raise the same kind of money with a less-specific project.   The MAUI crowd can better fund a project which specifically delivers a "Stride3D/Maui Render Surface" which also consumes Mouse/Keyboard inputs.  The deliverable must be usable "out of box".  Then we can get funding much easier.

The other target UIs (e.g. Avalonia) - can be put into separate bounties -- which then should be easier/cheaper, because the heavy-lifting will by then, already have been done.

Do we have any team members who would be willing to implement this for MAUI?  And if so, what is their "fixed price" bid?

(Is there a bidding process here, to help establish the bounty?)

**Reply by najak3d** [reply]
I put in $50 towards this feature just now, just to declare my official interest.  I can get considerably more once this feature is updated to include MAUI as the first platform to become supported.    I believe raising $1000 will be a cinch if the feature can be catered to deliver MAUI first as part of the bid/bounty.

**Reply by Aggror** [reply]
I am awaiting the response from @TheKeyblader who worked on this PR: https://github.com/stride3d/stride/pull/1315 

During the last community meeting, an amount for this particular PR was not set yet. If the developer can give an indication on progress, I can set a target goal for it, as well as a separate target for implementing the Maui approach. We would have to go over the details with other contributors first though.

**Reply by najak3d** [reply]
Thank you for the response.   We're not in a huge rush for this, as we're built soundly now upon UrhoSharp, but also realize that UrhoSharp has a shelf-life on it, since nobody in this world seems to be able to rebuild it or maintain it -- so it remains as "it is what it is, without any chance for bug fixes".   It works fine for us, but for the future, we're wanting something with life/hope, and maintainable/debuggable -- which will be Stride3D.    We're in this boat with other Xamarin Forms apps, without a good solution.

Microsoft folks (including Miguel De Icaza, and Egor Bogatov) have directly expressed an interest in seeing the MAUI ecosystem enriched by a good 3D rendering solution, and even spent some time trying (unsuccessfully) to revive UrhoSharp.   I believe raising $1000 for MAUI support should be a cinch, once we have the project well-defined with bid/commitment in place (pending the bounty).

**Reply by najak3d** [reply]
As of 2020, Microsoft advertised UrhoSharp on their own Xamarin Forms pages -- it was their primary recommendation for 3D rendering.   But when UrhoSharp was reported as unbuildable and not-maintained, and with some issues with no one to address -- then Microsoft removed UrhoSharp from their pages -- now leaving a void (i.e. no 3D rendering solution other than a raw OpenGL form).

I believe that Stride3D may actually become a MAUI advertised solution, direct from the Microsoft MAUI pages.  That's what I'll be pushing for.

**Comment by Nerdner** [comment]
Hello @Aggror — I found this discussion through Stride's bounty documentation. Are there any small, currently funded and unassigned C# bug investigations or fixes available? I would use AI assistance and provide a tested, reviewable contribution. Before starting, could you confirm a suitable issue, its acceptance criteria and bounty amount, and whether PayPal payout through Open Source Collective is still available? I am happy to keep payment details in the private payout process. Thank you.
